util/pullmetrics.py

```diff
@@ -0,0 +1,238 @@
+        app.extensions = getattr(app, "extensions", {})
+        app.extensions["pullmetrics"] = pull_metrics
+        return pull_metrics
+
+    def __getattr__(self, name):
+        return getattr(self.state, name, None)
+
+
+class PullMetrics(object):
+    """
+    Defines a helper class for tracking pull metrics as backed by Redis.
+    """
+
+    def __init__(self, redis_config):
+        self._redis = redis.StrictRedis(socket_connect_timeout=2, socket_timeout=2, **redis_config)
+
+    @staticmethod
+    def _tag_pull_key(repository, tag_name):
+        return "pull_events:repo:%s:tag:%s" % (repository, tag_name)
+
+    @staticmethod
+    def _manifest_pull_key(repository, manifest_digest):
+        return "pull_events:repo:%s:digest:%s" % (repository, manifest_digest)
+
+    def track_tag_pull_sync(self, repository_ref, tag_name, manifest_digest):
+        """
+        Synchronously track a tag pull event.
+
+        Args:
+            repository_ref: Repository object or repository_id
+            tag_name: Name of the tag
+            manifest_digest: Manifest digest
+        """
+        from data import model
+
+        # Get repository_id if object is passed
+        repository_id = repository_ref.id if hasattr(repository_ref, "id") else repository_ref
+        repository_path = (
+            f"{repository_ref.namespace_name}/{repository_ref.name}"
+            if hasattr(repository_ref, "namespace_name")
+            else str(repository_id)
+        )
+
+        timestamp = int(datetime.utcnow().timestamp())
+
+        # Create keys for tag and manifest
+        tag_key = self._tag_pull_key(repository_path, tag_name)
+        manifest_key = self._manifest_pull_key(repository_path, manifest_digest)
+
+        # Use Redis pipeline for atomic operations
+        pipe = self._redis.pipeline()
+
+        # Tag pull event - the worker will create both tag and manifest stats from this
+        pipe.hset(tag_key, "repository_id", repository_id)
+        pipe.hset(tag_key, "tag_name", tag_name)
+        pipe.hset(tag_key, "manifest_digest", manifest_digest)
+        pipe.hincrby(tag_key, "pull_count", 1)
+        pipe.hset(tag_key, "last_pull_timestamp", timestamp)
+        pipe.hset(tag_key, "pull_method", "tag")
+
+        pipe.execute()
+
+        logger.debug(
+            "Tracked tag pull: repo_id=%s tag=%s digest=%s",
+            repository_id,
+            tag_name,
+            manifest_digest,
+        )
+
+    def track_tag_pull(self, repository, tag_name, manifest_digest):
+        """
+        Track a tag pull event.
+
+        Note that this occurs in a thread to prevent blocking.
+        """
+
+        def conduct():
+            try:
+                self.track_tag_pull_sync(repository, tag_name, manifest_digest)
+            except redis.RedisError:
+                logger.exception("Could not track tag pull metrics")
+
+        thread = threading.Thread(target=conduct)
+        thread.start()
```

Comment:
I think we need to `.join()` these threads at some point to prevent leaking/to clean up the resources and free the thread. It may be good to try using a thread pool where we can also rate limit the amount of threads spawned in case of high traffic